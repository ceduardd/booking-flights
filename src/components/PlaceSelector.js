import { useState } from 'react';
import axios from 'axios';
import { useSelector } from 'react-redux';

const PlaceSelector = ({
  label,
  value,
  changeHandler,
  placeholder,
  menuInstruction,
  errorMessage,
}) => {
  const { errorFormat } = useSelector(state => state.ui);

  const [results, setResults] = useState([]);

  // Filter results of airports
  const filterHandler = async e => {
    changeHandler(e.target.value);

    if (e.target.value.trim().length > 2) {
      const { data } = await axios.get(
        `https://autocomplete.edestinos.com/?query=${e.target.value}&locale=en_US`
      );

      // Format payload
      const airports = data.result.filter(data => data._type === 'airport');

      const suggestions = airports.map(airport => ({
        code: airport._source.code,
        name: airport._source.suggestion,
        city: airport._source.cityName,
        country: airport._source.countryName,
      }));

      setResults(suggestions);
    }
  };

  // Set value on input
  const selectHandler = value => {
    changeHandler(value);

    setResults([]);
  };

  // Unselect dropdown list
  document.addEventListener(
    'click',
    e => {
      if (results.length !== 0) {
        setResults([]);
      }
    },
    false
  );

  return (
    <div className="selector__container">
      <label>
        {label}:
        <div style={{ marginTop: '5px' }} className="form-group">
          <input
            className="form-control"
            type="text"
            placeholder={placeholder}
            value={value}
            onChange={e => filterHandler(e)}
          />
          <i className="fas fa-map-marker-alt"></i>
        </div>
        {results.length !== 0 && (
          <div className="selector__menu" aria-hidden={true}>
            <span className="selector__item selector__item-instruction">
              {menuInstruction}
            </span>
            {results.map(result => (
              <span
                key={result.code}
                onClick={() => selectHandler(`${result.city} - ${result.code}`)}
                className="selector__item"
              >
                {result.name}
              </span>
            ))}
          </div>
        )}
      </label>
      {errorFormat && errorFormat === errorMessage && (
        <div className="error-label">
          <i className="fas fa-exclamation-circle"></i> {errorMessage}
        </div>
      )}
    </div>
  );
};

export default PlaceSelector;
