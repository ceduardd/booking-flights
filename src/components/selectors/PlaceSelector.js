import { useState } from 'react';
import { useSelector } from 'react-redux';
import { getSuggestions } from '../../helpers/getSuggestions';

import ErrorLabel from '../ErrorLabel';

const PlaceSelector = ({
  label,
  value,
  changeHandler,
  placeholder,
  menuInstruction,
  errorMessage,
}) => {
  const { errorFormat } = useSelector(state => state.ui);

  const [results, setResults] = useState([]);

  // Filter results of airports
  const filterHandler = async e => {
    changeHandler(e.target.value);

    if (e.target.value.trim().length > 2) {
      const suggestions = await getSuggestions(e.target.value);

      setResults([...suggestions]);
    }
  };

  // Set value on input
  const selectHandler = value => {
    changeHandler(value);

    setResults([]);
  };

  // Unselect dropdown list
  document.addEventListener(
    'click',
    e => {
      if (results.length !== 0) {
        setResults([]);
      }
    },
    false
  );

  return (
    <div className="selector__container">
      <label>
        <span>{label}:</span>
        <div style={{ marginTop: '5px' }} className="form-group">
          <input
            className="form-control"
            type="text"
            placeholder={placeholder}
            value={value}
            onChange={e => filterHandler(e)}
          />
          <i className="fas fa-map-marker-alt"></i>
        </div>

        {results.length !== 0 && (
          <div className="selector__menu">
            <span className="selector__item selector__item-instruction">
              {menuInstruction}
            </span>
            {results.map(result => (
              <span
                key={result.code}
                onClick={() => selectHandler(`${result.city} - ${result.code}`)}
                className="selector__item"
              >
                {result.name}
              </span>
            ))}
          </div>
        )}
      </label>
      {errorFormat && errorFormat === errorMessage && (
        <ErrorLabel>{errorMessage}</ErrorLabel>
      )}
    </div>
  );
};

export default PlaceSelector;
