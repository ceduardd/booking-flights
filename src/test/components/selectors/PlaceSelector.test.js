import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import PlaceSelector from '../../../components/selectors/PlaceSelector';
import configureStore from 'redux-mock-store';

const mockStore = configureStore();

const initialState = {
  ui: {},
};

let store = mockStore(initialState);

store.dispatch = jest.fn();

const changeHandler = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <PlaceSelector changeHandler={changeHandler} />
  </Provider>
);

describe('<PlaceSelector />', () => {
  beforeEach(() => {
    store = mockStore(initialState);
    jest.clearAllMocks();
  });

  test('should be match with snapshot ', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should call to changeHandler', () => {
    const input = wrapper.find('.form-control');

    const e = {
      target: {
        value: 'madrid',
      },
    };

    input.simulate('change', e);

    expect(changeHandler).toHaveBeenCalledWith(e.target.value);
  });

  test('should show error message', () => {
    const errorMessage = 'Something error';

    const state = {
      ui: {
        errorFormat: errorMessage,
      },
    };

    const store = mockStore(state);

    const wrapper = mount(
      <Provider store={store}>
        <PlaceSelector errorMessage={errorMessage} />
      </Provider>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('ErrorLabel').exists()).toBe(true);
  });
});
