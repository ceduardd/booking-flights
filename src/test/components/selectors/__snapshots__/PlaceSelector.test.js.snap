// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlaceSelector /> should be match with snapshot  1`] = `
<div
  className="selector__container"
>
  <label>
    <span>
      :
    </span>
    <div
      className="form-group"
      style={
        Object {
          "marginTop": "5px",
        }
      }
    >
      <input
        className="form-control"
        onChange={[Function]}
        type="text"
      />
      <i
        className="fas fa-map-marker-alt"
      />
    </div>
  </label>
</div>
`;

exports[`<PlaceSelector /> should show error message 1`] = `
<div
  className="selector__container"
>
  <label>
    <span>
      :
    </span>
    <div
      className="form-group"
      style={
        Object {
          "marginTop": "5px",
        }
      }
    >
      <input
        className="form-control"
        onChange={[Function]}
        type="text"
      />
      <i
        className="fas fa-map-marker-alt"
      />
    </div>
  </label>
  <div
    className="error-label"
  >
    <i
      className="fas fa-exclamation-circle"
    />
     
    Something error
  </div>
</div>
`;
